### PR TITLE
fix(release): authenticate fragment cleanup apply step and report retention trend (#2898)

### LINKED ISSUE #2898: [Workstream I] PR 1 of 13: Authenticate the release fragment cleanup apply step, guard gh-dependent steps, and report retention days-to-limit
**Roadmap:** I1 · **Workstream I: Frontier-audit remediation and enhancement (2026-09-21)**
**Sequence:** PR 1 of 13 (tracking slot; dependencies define execution order)
**Roadmap baseline:** 2026-09-22 · **Priority:** P0 · **Repo revision:** 006694310

## Outcome

Every release-please release runs the documented fragment cleanup to completion: `docs/releases/v<version>.md` and `docs/releases/manifests/v<version>.json` are materialized and a cleanup PR is opened, and a workflow-shape test makes it impossible to ship a `release-notes-fragments.mjs` step that calls `gh` without a token again. `verify-retention` reports days-to-limit so the next retention wall is visible weeks ahead.

## Problem

The `cleanup-release-fragments` job has failed on every release since it was introduced on 2026-09-06, and nothing surfaces it because `publish-npm` does not depend on it.

- The "Validate and apply fragment cleanup" step ([release-and-publish.yml:234-243](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/.github/workflows/release-and-publish.yml#L234-L243)) sets only `TAG_NAME` in `env:`; there is no job-level `env:` (job keys at [:209-243](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/.github/workflows/release-and-publish.yml#L209-L243)) and the checkout uses `persist-credentials: false` (:224-228). The very next step ("Open cleanup PR") carries `GH_TOKEN` and `GITHUB_REPOSITORY` ([:245-248](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/.github/workflows/release-and-publish.yml#L245-L248)).
- `apply-cleanup` → `modeApplyCleanup` → `refreshCleanupPlanProvenance` ([release-notes-fragments.mjs:2195-2197](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/scripts/release-notes-fragments.mjs#L2195-L2197)) → `peelRemoteTagObject` → `ghJson` (`execFileSync('gh', …)`, [:1116](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/scripts/release-notes-fragments.mjs#L1116)).
- Live failure (run 35585492871, job 106287917782, v7.185.0): `[release-notes-fragments] ERROR: Error: Command failed: gh api repos/ZaxbyHub/opencode-swarm/git/ref/tags/v7.185.0` / `gh: To use GitHub CLI in a GitHub Actions workflow, set the GH_TOKEN environment variable` at `refreshCleanupPlanProvenance (…mjs:2197:23)` / `modeApplyCleanup (…mjs:2232:16)`. Every release-please run inspected 2026-09-12..09-21 shows `cleanup-release-fragments = failure`, `update-release-notes = success`, `publish-npm = success`; a non-release run shows the job `skipped`.
- `git blame` → step introduced by 24296284f (2026-09-06, "feat(release): automate safe fragment reconciliation"); the `gh` dependency in the apply path by 0a7eb4001 (2026-09-06, "fix(release): harden cleanup provenance and CI gates"). Last materialized archive file: `docs/releases/v7.166.4.md` (same commit). Nothing above v7.166.4 exists in `docs/releases/` or `docs/releases/manifests/` although `CHANGELOG.md` is at 7.185.0.
- Compounding wall: `MAX_PENDING_FRAGMENT_RETENTION = 750` ([mjs:423](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/scripts/release-notes-fragments.mjs#L423)) is the default `maxPendingFragments` for `auditFragmentRetention` (:621-623) which `verify-retention` runs from [drift-check.yml:67-68](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/.github/workflows/drift-check.yml#L67-L68) on every pull_request/push/merge_group. `ls docs/releases/pending | wc -l` = 502 on 2026-09-21; September adds 4–11/day. Projected limit ~2026-10-28. `drift` is `intendedRequiredContexts` only ([required-check-contract.json:25](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/scripts/required-check-contract.json#L25)), so it reddens without blocking. The historical-replay suppression hatch requires a committed, in-TTL replay state and turns into a violation when stale (`:689-726`, `:868-873`), so it is not an escape.
- Documented contract: [contributing.md:405-459](https://github.com/ZaxbyHub/opencode-swarm/blob/00669431030c350a5c28c41710f30e93fe3f40ca/contributing.md#L405-L459) describes the tag job writing the archive and opening a deterministic cleanup PR in the present tense.

Provenance: audit finding D3 (reviewer UPHELD, reproduced live; critic SURVIVES). No prior issue mentions the job, `GH_TOKEN` in the apply step, or the archive freeze (zaxbygraph search over 2,892 items).

## Requested change

Make the apply step authenticated, make the class impossible to reintroduce, and make the retention trend visible.

## Prerequisites and ownership

Depends on nothing. OWNS `.github/workflows/release-and-publish.yml` step env, the workflow-shape test, and the `verify-retention` diagnostics. Does NOT own the historical back-fill or the frontmatter rendering (I2), nor the required-check promotion of `drift`. Parallel-safe with I3–I13. **Must merge before I2.**

## Required scope

- [ ] Add `GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}` and `GITHUB_REPOSITORY: ${{ github.repository }}` to the "Validate and apply fragment cleanup" step (match the next step's env exactly).
- [ ] Workflow-shape test (`tests/security/` or `tests/unit/scripts/ci/`): for every `run:` step in `release-and-publish.yml` and `drift-check.yml` that invokes `release-notes-fragments.mjs` with a mode that calls `ghJson` (`update-pr`, `update-release`, `prepare-cleanup`, `apply-cleanup`), assert the step env carries `GH_TOKEN`. Derive the gh-dependent mode list from the script (export a `MODES_REQUIRING_GH` constant) rather than hardcoding it in the test.
- [ ] `verify-retention` prints `pending`, `limit`, the 14-day add rate from `git log --diff-filter=A`, and projected days-to-limit; emit a `::warning::` when < 30 days.
- [ ] Add the cleanup job to the drift sticky PR comment or a post-release health line so a failing cleanup is visible without opening the run (minimal form acceptable: a `::error::` annotation in the job plus a `needs:`-independent summary step that reports the cleanup conclusion).
- [ ] Release fragment under `docs/releases/pending/`.
- [ ] Out of scope: back-filling v7.166.5..v7.185.x (I2), frontmatter stripping (I2), promoting `drift` to required.

## Acceptance

| Case | Required proof |
|---|---|
| Apply step authenticated | Next release run: `cleanup-release-fragments = success`; `docs/releases/v<version>.md` + manifest appear on the cleanup PR |
| Shape test guards the class | Test fails when `GH_TOKEN` is removed from any gh-dependent step (show the red run or local failure output) |
| Retention trend visible | `node scripts/release-notes-fragments.mjs verify-retention` output shows pending/limit/rate/days-to-limit |
| Failure is visible | Cleanup failure produces an annotation or summary line without opening the job log |
| No regression | `bun run check:pre-push`, `bun run drift:check --enforce`, quality job green |

## Evidence and exit gate

Attach: the green cleanup job URL for the first post-merge release, the shape-test red/green pair, `verify-retention` output, `git diff` of the workflow. Invalidated if the release workflow is restructured before this lands.

## Execution and closure contract

Treat PR 1 of 13 as a stable tracking slot, not merge order. Re-anchor source and active-PR overlap before implementation. Close only the fully satisfied scope: wired production callers, meaningful focused validation and documented outcomes, no silent deferral. Open one PR per issue titled with this issue's prefix; add tests for every acceptance row; update docs touched by the change; ship a pending release fragment.

## Complete workstream execution index

Slots identify work; they do not override dependencies. Totals are authoritative for Workstream I (baseline 2026-09-22). Epic: #2897. Audit artifacts: `.swarm/frontier-audit/20260921T120000Z/REPORT.md` (audit worktree; gitignored).

### Workstream I: Frontier-audit remediation and enhancement (13 slots)

| Slot | Issue | Work |
|---|---|---|
| I1 | #2898 | Authenticate the release fragment cleanup apply step, guard gh-dependent steps, and report retention days-to-limit |
| I2 | #2899 | Back-fill the docs/releases archive v7.166.5..current and stop frontmatter leaking into release bodies without breaking the provenance oracle |
| I3 | #2900 | Renew or retire the quarantine entries that hard-fail the required quality job from 2026-10-11 |
| I4 | #2901 | Wire or remove the inert turbo_mode key and fix the parallelization sub-key promises and doctor advisory |
| I5 | #2902 | Host-contract drift check that compares the #2526 fixture against the OpenCode host source at npm-latest |
| I6 | #2903 | Diff-scoped ratchet for vacuous test assertions and cleanup of the 15 structurally cannot-fail sites |
| I7 | #2904 | Config-consumption ratchet: every PluginConfigSchema key declares a consumer or an inert marker; doctor warns on inert keys |
| I8 | #2905 | Expiry-aware quarantine census, renewal-requires-issue policy, and a weekly aging workflow |
| I9 | #2906 | Counterbalanced candidate order across reviewer passes with per-finding order-agreement on verdict receipts |
| I10 | #2907 | Align shipped skills with the Agent Skills specification: metadata map for non-spec keys and a 500-line body budget |
| I11 | #2908 | Settle/admit authority census with a ratchet, and property tests that make the four settlement-wedge shapes unreachable |
| I12 | #2909 | ADR-0003 and hook inventory for the OpenCode 2 plugin API, with a dual-shape server()/setup() spike on both hosts |
| I13 | #2910 | Port guidance, tools, hooks, and lifecycle to the OpenCode 2 plugin API behind a dual-host CI lane |

Dependencies: I2 after I1 · I7 after I4 · I8 after I3 · I13 after I12. All other slots are parallel-safe.

### PR DESCRIPTION
**Pull request by zaxbysauce**
Closes #2898

## Summary

The `cleanup-release-fragments` job has failed on every release since it was introduced on 2026-09-06: the "Validate and apply fragment cleanup" step ran `release-notes-fragments.mjs apply-cleanup` with only `TAG_NAME` in its step env, so the script's first `gh api` call (via `refreshCleanupPlanProvenance` → `peelRemoteTagObject` → `ghJson`) exited with `gh: To use GitHub CLI in a GitHub Actions workflow, set the GH_TOKEN environment variable`. Because `publish-npm` does not depend on the cleanup job, npm publishes stayed green, the failure was invisible, and the `docs/releases/` archive froze at v7.166.4 (live evidence: run 35585492871 / job 106287917782 for v7.185.0).

This PR closes the defect at three levels:

- **Instance** — the apply step now carries `GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}` and `GITHUB_REPOSITORY: ${{ github.repository }}`, byte-matching the adjacent "Open cleanup PR" step.
- **Class** — `release-notes-fragments.mjs` exports `MODES_REQUIRING_GH` (exactly the four CLI modes whose execution path reaches `gh`: `update-pr`, `update-release`, `prepare-cleanup`, `apply-cleanup`), and a new workflow-shape test (`tests/unit/scripts/ci/release-fragments-gh-auth-shape-2898.test.ts`) scans `release-and-publish.yml` and `drift-check.yml` and fails naming the job/step/modes if any step invoking a gh-dependent mode lacks `GH_TOKEN` in its effective (job ∪ step) env. Reintroducing the defect class is now a unit-shard failure, proven by a mutation probe (stripping `GH_TOKEN` from the workflow flips the test red).
- **Visibility + trend** — `verify-retention` now prints the 14-day add rate (from `git log --diff-filter=A`) and projected days-to-limit, with a `::warning::` under 30 days (live output at head: `add rate 6.79/day (14-day window, 95 added); days-to-limit 35.8`); and a new `release-health-summary` job runs on every release regardless of the cleanup result (`if: always()`), reporting job conclusions and emitting an `::error::` annotation when the cleanup job did not complete — a failing cleanup is visible without opening the job log, without coupling `publish-npm` to the cleanup path.

Out of scope (per the issue): the v7.166.5..current archive back-fill and render-only frontmatter handling (I2 / #2899), and promoting `drift` to a required check.

## Invariant audit

- 1 (plugin init): not touched — no `src/` change; `node scripts/repro-704.mjs` PASS (T3 server() 23.8ms).
- 2 (runtime portability): not touched — no bundle/export/entry change; `bun run build` PASS + `node --input-type=module -e "await import('./dist/index.js')"` PASS.
- 3 (subprocesses): touched — one new bounded git invocation (`git log --diff-filter=A --name-only --pretty=format: --since <iso> -- docs/releases/pending`) added to `scripts/release-notes-fragments.mjs`, routed through the existing `gitText` runner (explicit cwd, 15 s timeout, ignore-stdin, bounded buffer); no new spawn forms anywhere. `check:bare-spawn` PASS.
- 4 (.swarm containment): not touched — no runtime state writes; script writes only stdout and the pre-existing docs/releases flows.
- 5 (plan durability): not touched — no plan/ledger surface.
- 6 (test_runner safety): not touched.
- 7 (test writing): touched — two new bun:test files (7 + 9 tests), DI-injected git runner (no `mock.module`, no temp dirs), both under the 500-line cap; `check:mock-cleanup`, `check:test-tmpdir`, `check:test-file-cap` PASS.
- 8 (session state): not touched.
- 9 (guardrails/retry): not touched.
- 10 (chat/system msg): not touched.
- 11 (tool registration): not touched — `check-tool-registration` PASS, `bun run drift:check --enforce` clean for tools/skills/commands.
- 12 (release/cache): touched — this IS the release pipeline fix: the apply step env, the new summary job, and the retention diagnostics. Pending release fragment `docs/releases/pending/2898-cleanup-apply-gh-auth-retention.md` shipped (`check:pending-fragment` PASS). No `package.json#version`/`CHANGELOG.md`/manifest edits.

## Test plan

- Frozen acceptance checks C1–C6 (issue-tracer red checkpoint, anchor published on the issue): all RED→GREEN / GREEN→GREEN at head 49d75258c (`repro-check.sh run`, verdict PASS ×6; `verify-checkpoint` + `verify-anchor` PASS).
- `bun test tests/unit/scripts/ci/release-fragments-gh-auth-shape-2898.test.ts tests/unit/scripts/release-notes-fragments-retention-trend.test.ts` → 16 pass / 0 fail.
- Existing fragment suites (12 files, 204 tests incl. `release-fragment-cleanup-workflow.test.ts` pins) → 0 fail.
- Mutation probe: `GH_TOKEN` stripped from the apply step in a working copy → shape test 2 fail; restored → 7 pass; tree clean after.
- Quality battery on the committed tree: typecheck, tool-registration, mock-cleanup, invariants, cross-contamination, test-clock, runtime-src-refs, events, retention, registry-citations, core-events, shell-audit, trajectory-store, test-file-cap, pending-fragment, gate-portability, bare-spawn, test-tmpdir, bash-portability, error-channel-discard, path-identity, token-formula → ALL PASS; biome (CI-equivalent scoped, pinned 2.3.14) → exit 0.
- AC6 gate evidence at this exact head SHA (PR CI, Linux LF checkouts): `drift` = success — https://github.com/ZaxbyHub/opencode-swarm/actions/runs/35705396849 (runs `scripts/drift-check.ts --enforce` + `verify-retention`; its log shows the new trend line live: `retention trend: add rate 6.86/day (14-day window, 96 added); days-to-limit 35.4`) and `quality` = pass — https://github.com/ZaxbyHub/opencode-swarm/actions/runs/35705396861/job/106675319594 (ratchet battery). Both bound to head_sha 49d75258cf85172c464ca31de1c5830defec99ca. On the Windows dev worktree the same two commands exit 1 with `WORKFLOW_CHANGED_AFTER_CAPTURE` on `pr-standards.yml` — a CRLF-checkout artifact that reproduces identically at the base commit (detached-worktree control) and whose hash-captured files show an empty `git diff base..HEAD`; the CI runs above are the authoritative AC6 result.
- Post-merge (per the issue's exit gate, attached to the issue on the first release after merge): `cleanup-release-fragments = success` with `docs/releases/v<version>.md` + manifest on the cleanup PR.

## Acceptance Criteria -> Evidence

| Acceptance criterion (from intake) | Evidence |
|---|---|
| AC1 apply step authenticated | Frozen C1 RED at base → GREEN at head (repro-check.sh run, log repro/C1.{base,head}.log); yaml env parity verified in review; workflow diff in this PR |
| AC2 shape test guards the class | Frozen C2 RED→GREEN; tests/unit/scripts/ci/release-fragments-gh-auth-shape-2898.test.ts 7/7 pass; mutation pair: GH_TOKEN stripped from the workflow → 2 tests fail, restored → pass |
| AC3 retention trend visible | Frozen C3 RED→GREEN; local run and CI drift job log: `retention trend: add rate 6.86/day (14-day window, 96 added); days-to-limit 35.4`; trend unit suite 9/9 |
| AC4 failure visible | Frozen C4 RED→GREEN; release-health-summary job with `if: always()` reports cleanup conclusion and emits `::error::` on non-success/non-skipped |
| AC5 release fragment shipped | docs/releases/pending/2898-cleanup-apply-gh-auth-retention.md (check:pending-fragment PASS) |
| AC6 no regression | Exact-SHA PR CI: drift = success (run 35705396849), quality = pass (job 106675319594), both bound to head 49d75258c; preserving suites C6 GREEN→GREEN; full matrix completing on the PR |
| Post-merge row (issue exit gate) | First release run after merge: cleanup-release-fragments = success + materialized docs/releases/v<version>.md on the cleanup PR — to be attached to issue #2898 when it exists |

## Waivers (or none)

None. Every Full-Resolution Contract clause is satisfied with evidence; no waiver invoked.

Merge status: AWAITING_USER_APPROVAL (merge is separately gated; this PR will not be merged by the implementer without explicit recorded approval).

PR: https://github.com/ZaxbyHub/opencode-swarm/pull/2911
PR head: 49d75258cf85172c464ca31de1c5830defec99ca

### CODE REVIEW

**Comment by zaxbysauce** [comment]
## swarm-pr-review — PR #2911

**Verdict: `APPROVE`**

### PR intent

Closes #2898. Three-level fix for the `cleanup-release-fragments` release job, which had failed on every release since 2026-09-06: (1) the "Validate and apply fragment cleanup" step now carries `GH_TOKEN`/`GITHUB_REPOSITORY`, matching its sibling "Open cleanup PR" step; (2) `MODES_REQUIRING_GH` is exported and a new workflow-shape test makes the missing-token defect class impossible to reintroduce without a red unit shard; (3) `verify-retention` now reports a 14-day add-rate trend with a days-to-limit projection (`::warning::` under 30 days), and a new `release-health-summary` job (`if: always()`) surfaces a non-success cleanup conclusion with an `::error::` annotation, without coupling `publish-npm` to the cleanup path.

### Implementation summary

- `.github/workflows/release-and-publish.yml`: +2 env lines on the apply-cleanup step; new `release-health-summary` job (25 lines).
- `scripts/release-notes-fragments.mjs`: `modeVerifyRetention` extended with trend computation/logging; new exported `MODES_REQUIRING_GH`, `RETENTION_TREND_WINDOW_DAYS`, `RETENTION_WARN_DAYS`, `computeRetentionTrend`, `countPendingAddsSince`.
- Two new test files (201 + 158 lines, both under the FR-006 500-line cap): `tests/unit/scripts/ci/release-fragments-gh-auth-shape-2898.test.ts`, `tests/unit/scripts/release-notes-fragments-retention-trend.test.ts`.
- `contributing.md` and `docs/releases/pending/2898-cleanup-apply-gh-auth-retention.md` updated/added.

### Intended vs actual mapping

| Obligation | Source | Actual evidence | Status |
|---|---|---|---|
| AC1 apply step authenticated | PR body | `release-and-publish.yml:236-238` env block byte-matches the adjacent "Open cleanup PR" step (independently re-verified) | MET |
| AC2 shape test guards the class | PR body | `release-fragments-gh-auth-shape-2898.test.ts`: 7/7 pass baseline; mutation probe (GH_TOKEN stripped) independently reproduced -> 2 fail, restored -> 7 pass, tree clean | MET |
| AC3 retention trend visible | PR body | `computeRetentionTrend`/`countPendingAddsSince` unit-tested incl. null-trend, over-limit, and ISO-8601-validation branches; math re-derived by hand (rounding, div-by-zero guards) | MET |
| AC4 failure visible | PR body | `release-health-summary` job: correct needs/`if: always()` wiring, does not add an edge to `publish-npm`, bash conditional correctly distinguishes success/skipped from failure/cancelled/timed_out | MET |
| AC5 release fragment shipped | PR body | `docs/releases/pending/2898-cleanup-apply-gh-auth-retention.md` present, accurate, matches sibling-fragment conventions | MET |
| AC6 no regression | PR body | Full 7-lane sweep found zero regressions; `verify-retention`'s added `trend` JSON field has no consumers in-repo (drift-check.yml runs it without piping) | MET |
| Closes #2898 | Issue linkage | Issue #2898 is OPEN; diff addresses every stated acceptance criterion; root cause (missing job/step env) matches the issue's own live-failure evidence | MET |

### Validation provenance

- **Capability profile:** B (Claude Code native parallel subagents, no swarm-plugin controller). Depth tier: **M** (541 changed lines across 6 files; risk triggers: CI/release workflow change, new subprocess invocation).
- **Pre-flight:** PR head `49d75258c` fetched, checked out detached, and verified byte-identical to `gh pr view --json headRefOid` throughout the review. Merge base `76fa89125` computed against a freshly-fetched `origin/main`. Working tree clean at every checkpoint.
- **Phase 0A intake:** No prior human/bot review comments or requested changes exist on this PR (`pulls/2911/reviews` and `pulls/2911/comments` both empty). One informational drift-check bot notice (required-check-contract ruleset-divergence notices, non-blocking, unrelated to this PR's changes).
- **Mergeability:** `MERGEABLE` / `mergeStateStatus: CLEAN`.
- **Base dimensions covered with attestation:** 6/6 (consolidated into 4 lanes per tier-M dispatch rules: intent-architecture+compatibility-delivery, correctness-state+reliability-performance, tests-falsifiability, security-trust).
- **Micro risk families evaluated and attested:** all 11 triaged; 6 MATCHED and dispatched (auth-identity-secrets, subprocess-platform, dependencies-build-release, api-schema-migrations, test-infrastructure, unclassified-risk fallback), 5 NOT_TRIGGERED with absence evidence (untrusted-input-boundaries - only trusted local git-log output parsed; concurrency-state - no new shared/async state; ui-accessibility-i18n - no UI surface touched; privacy-observability - no telemetry/PII; generated-provenance - no generated/vendored artifacts).
- **Explorer lanes launched/returned:** 7/7, all with parseable CANDIDATE/CLEAN rows on first dispatch (no retries needed).
- **Raw candidates:** 1 (SEC-1, LOW).
- **Reviewer lanes dispatched/returned:** 1/1, independent re-derivation from source (permissions-block diff against base, full gh-call-site reachability trace from `modeApplyCleanup`), not a re-statement of the explorer's claim.
- **Findings confirmed by reviewer:** 0. Rejected as false positive: 0. Marked PRE_EXISTING: 1 (SEC-1). Left UNVERIFIED: 0.
- **Findings escalated to critic:** 0 - SEC-1 resolved to PRE_EXISTING/NONE severity, not HIGH/CRITICAL nor borderline MEDIUM, so critic dispatch was correctly skipped per protocol.
- **Falsification probes:** SEC-1's reviewer substituted two structural falsification attempts for a runtime probe (base-vs-head diff of the exact permissions block; exhaustive enumeration of every gh-call site reachable from `modeApplyCleanup`) - both failed to falsify the "not exploitable" conclusion, and no live-Actions-job probe is available in this environment.
- **Suppressed candidates:** none - the single candidate that survived explorer generation was routed to reviewer in full.

### Confirmed findings

None. **No confirmed findings — all validated lanes CLEAN.**

Lane-by-lane clean summary:
- intent-architecture + compatibility-delivery: obligations traced end-to-end, `MODES_REQUIRING_GH` confirmed consumed (not dead-exported), `release-health-summary` wiring confirmed non-blocking to `publish-npm`, docs match code.
- correctness-state + reliability-performance: trend math re-verified by hand (rounding, div-by-zero/clamp guards, UTC round-trip), bash conditional confirmed exhaustive over GHA job-result values, subprocess bounds (15s timeout, 16MB buffer) confirmed applied.
- tests-falsifiability: 16/16 tests independently re-run and passing; PR's own mutation-probe claim ("GH_TOKEN stripped -> 2 fail") independently reproduced exactly; tree restored clean.
- subprocess-platform: sole new subprocess call (`countPendingAddsSince`'s `git log`) uses array-argv execFileSync, bounded timeout/buffer, path-contained repoRoot, non-interactive stdio.
- dependencies-build-release + api-schema-migrations: no new blocking job edges; `verify-retention`'s added `trend` JSON field has zero in-repo consumers (backward compatible).
- test-infrastructure + unclassified-risk: no cross-platform/isolation issues (CI matrix is ubuntu-only for this suite); release fragment and docs accurate; no stray TODO/placeholder text.

### Pre-existing findings

One candidate (SEC-1) reduced to this category by independent reviewer re-verification — reported here for disclosure, not as an action item:

> The `cleanup-release-fragments` job's `permissions:` block (`actions: write`, `contents: write`, `pull-requests: write`, .github/workflows/release-and-publish.yml:217-220) is broader than the newly-authenticated "Validate and apply fragment cleanup" step strictly needs. The reviewer confirmed this permissions block is **byte-identical at the merge base** (not introduced by this PR — GitHub Actions has no step-level permission scoping, only job-level, and the job's other step already legitimately needs the full grant). More importantly, the reviewer traced every `gh` call reachable from `apply-cleanup` (`peelRemoteTagObject`, `refreshCleanupPlanProvenance`) and found the reachable surface is exactly **two read-only API calls**; the write-capable calls (`pr edit`, `release edit`) live in the separate `update-pr`/`update-release` code paths, not `apply-cleanup`. All `gh`/`git` invocations use array-argv `execFileSync` (no shell interpolation), and the workflow triggers only on `push: main`. **Not actionable** — flagged only for completeness per the protocol's disclosure requirement.

### Test / coverage gaps

None identified beyond what the PR itself already covers well (the new test files exercise the null-trend, over-limit, and ISO-8601-validation edge cases, plus a real mutation-probe pair for the shape test).

### Disproved candidates and false positives

None — SEC-1 was not disproved (its factual claims all held up), it was downgraded to PRE_EXISTING/NONE-severity/not-actionable after independent verification showed the excess scope has no reachable write path from the newly-authenticated step.

### Merge recommendation

**APPROVE.** All 7 stated acceptance criteria are met with independently-reproduced evidence (not just re-stated PR-body claims), the one candidate finding an explorer lane surfaced was independently re-verified and shown to be pre-existing and non-actionable, and the change carries no regression risk to `verify-retention`'s JSON contract or to `publish-npm`'s dependency graph. No action required before merge.

---
*Full swarm-pr-review: Profile B (Claude Code), depth tier M — 6/6 base dimensions + 6/6 matched micro-lanes explored, 1 candidate independently reviewed, 0 confirmed findings, 0 critic escalations needed.*

**Comment by zaxbysauce** [comment]
## swarm-pr-review — PR #2911 (round 2, run pr2911-20260922)

**Verdict: `APPROVE`** — zero CRITICAL/HIGH findings survive validation. Proceeding to feedback round per author instruction.

**Ladder:** 6 base explorer lanes + 4 consolidated micro-lanes attesting all 11 risk families (8 MATCHED, 3 NOT_TRIGGERED with absence evidence: concurrency-state, ui-accessibility-i18n, generated-provenance) → 19 raw candidates → 2 independent reviewer shards → 16 VERIFIED / 3 REJECTED → critic challenge on the 3 HIGHs (all downgraded: PRR-001's CI-impact claim disproved — `GITHUB_REPOSITORY` is a runner-injected default env var, so only `GH_TOKEN` can actually break CI; PRR-002 → MEDIUM coverage debt; PRR-010 → LOW doc nit). Provenance note: the pinned reviewer/critic models were unavailable (provider failures); the skill-sanctioned general-purpose fallback ran the validation passes.

**Will fix in the feedback round (FB-1..FB-6):**
| ID | Sev | Finding |
|---|---|---|
| FB-1 / PRR-010 | LOW | contributing.md:411-413 dangling "must" (also Copilot's inline) |
| FB-2 / PRR-002 | MEDIUM | `modeVerifyRetention` wrapper has zero direct coverage (warning branch, n/a text, exit contract) |
| FB-3 / PRR-001 | LOW | shape test asserts only GH_TOKEN; add the GITHUB_REPOSITORY assertion for depth |
| FB-4 / PRR-007 | MEDIUM | mode regex misses a backslash-newline YAML continuation between script name and mode |
| FB-5 / PRR-006 | MEDIUM | zero add-rate near the wall (e.g. 740/750) emits no warning — advisory misleadingly optimistic |
| FB-6 / PRR-013 | LOW | GUARDED_WORKFLOWS hardcoded; derive the scan from .github/workflows/*.yml |

**Dispositioned without change** (rationales in the handoff): 15s git-log timeout has ~90x measured headroom; cancelled-cleanup red badge is documented intended behavior; title-less fragment frontmatter has 12 sibling precedents and no tooling requirement; rounding assertions brittle-only; job-level token scope is an Actions platform limit; ref-scope bias is pessimistic-only; regex mention-vs-invocation latent fail-noisy; per-step mode dedupe cosmetic; RangeError unreachable; runner-argv equality is load-bearing (positive).

The earlier APPROVE comment on this PR predates this run; this round's validated findings are the authoritative set.

File: contributing.md [hunk 1]
```diff
@@ -407,7 +407,11 @@ After your PR merges, release-please opens or updates its release PR. CI runs `s
 The tag job then prepares an exact-tag cleanup plan. It proves the remote
 peeled tag, local peeled tag, and checkout HEAD are the same commit, binds the
 full GitHub Release body and each consumed fragment's SHA-256, and passes that
-plan to a fresh `main` checkout. The apply step is dry-run first and writes
+plan to a fresh `main` checkout. The apply step runs with
+`GH_TOKEN`/`GITHUB_REPOSITORY` step env (any step invoking a gh-dependent
+fragment-script mode must — enforced by
```
[the remark this answers is not in the document]
Comment (zaxbysauce) [reply contributing.md]: Fixed in 5e9ed8c2: the clause now reads "any step invoking a gh-dependent fragment-script mode must carry them — enforced by ...". The swarm-pr-review round also confirmed the underlying contract is enforced by tests/unit/scripts/ci/release-fragments-gh-auth-shape-2898.test.ts (now asserting both GH_TOKEN and GITHUB_REPOSITORY across all workflows).

**Comment by zaxbysauce** [comment]
## swarm-pr-feedback closure — PR #2911 (run pr2911-20260922)

All feedback resolved at head `5e9ed8c206731dc04b193396e9522084593a28f3` (reviewer OVERALL APPROVE + separate final-critic APPROVE on the delta):

| id | source | outcome | fix-ref | evidence |
|----|--------|---------|---------|----------|
| FB-1 / PRR-010 / Copilot inline | review + PR comment | FIXED | 5e9ed8c2 | contributing.md clause now "must carry them — enforced by …"; thread replied |
| FB-2 / PRR-002 | review (MEDIUM) | FIXED | 5e9ed8c2 | new CLI smoke test drives `verify-retention` end to end (exit 0, trend keys, diagnostics) |
| FB-3 / PRR-001 | review (downgraded LOW) | FIXED | 5e9ed8c2 | shape test requires GH_TOKEN **and** GITHUB_REPOSITORY; strip-GITHUB_REPOSITORY mutation probe added |
| FB-4 / PRR-007 | review (MEDIUM) | FIXED | 5e9ed8c2 | separator class `[\s\]+` tolerates backslash-newline continuations; continuation-form mutation probe added |
| FB-5 / PRR-006 | review (MEDIUM) | FIXED | 5e9ed8c2 | `retentionWallWarning` (unit-tested, boundary 675/674) warns when pending ≥ 90% of limit with a stalled add rate |
| FB-6 / PRR-013 | review (LOW) | FIXED | 5e9ed8c2 | workflow scan derived from `readdirSync(.github/workflows/*.yml)` — no hardcoded allowlist |
| PRR-003 | review | DISPOSITIONED | — | 15s git-log timeout has ~90x measured headroom (0.16s actual) |
| PRR-005 / 009 / 011 | review | REJECTED by reviewers | — | cancelled-red is documented intent; conclusion-flip omission not misleading; title-less fragments have 12 sibling precedents |
| PRR-004 / 012 / 014–018 | review | DISPOSITIONED | — | latent/brittle-only/platform-limited/pessimistic-bias/cosmetic/unreachable; rationales in the review comment |

CI: 45/45 green at the shipped head. Merge enqueued via the merge queue (user-authorized).
